File: packages/domain/src/value-objects/sku.ts
import { InvalidSkuError } from '../errors/domain-error.js';

/** Stock-keeping unit identifier for a Product. */
export class Sku {
  private constructor(readonly value: string) {}

  static parse(value: string): Sku {
    const trimmed = value.trim();
    if (!/^[A-Za-z0-9._-]{1,64}$/.test(trimmed)) {
      throw new InvalidSkuError();
    }
    return new Sku(trimmed);
  }

  toString(): string {
    return this.value;
  }
}

File: backend/crates/domain-inventory/src/sku.rs
use crate::error::InventoryError;

/// Stock-keeping unit identifier for a Product (GLOSSARY: Sku).
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Sku(String);

impl Sku {
    pub fn parse(value: &str) -> Result<Self, InventoryError> {
        let trimmed = value.trim();
        if trimmed.is_empty()
            || !trimmed
                .bytes()
                .all(|b| b.is_ascii_alphanumeric() || b == b'-' || b == b'_')
        {
            return Err(InventoryError::InvalidSku);
        }
        Ok(Self(trimmed.to_owned()))
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}
